vpr/src/place/analytical_placement/AnalyticalSolver.cpp

```diff
@@ -1,29 +1,74 @@
 #include "AnalyticalSolver.h"
+#include <Eigen/src/Cholesky/LDLT.h>
+#include <Eigen/src/Core/util/Constants.h>
+#include <Eigen/src/SparseCholesky/SimplicialCholesky.h>
+#include <Eigen/src/SparseCore/SparseMatrix.h>
 #include <Eigen/Sparse>
+#include <Eigen/Eigenvalues>
 #include <Eigen/IterativeLinearSolvers>
+#include <Eigen/Cholesky> 
+#include <cstdio>
 #include <memory>
+#include "PartialPlacement.h"
 #include "atom_netlist.h"
 #include "globals.h"
+#include "partition_region.h"
 #include "vpr_context.h"
 #include "vpr_error.h"
 #include "vtr_assert.h"
 
-std::unique_ptr<AnalyticalSolver> make_analytical_solver(e_analytical_solver solver_type) {
+std::unique_ptr<AnalyticalSolver> make_analytical_solver(e_analytical_solver solver_type, PartialPlacement &p_placement) {
     if (solver_type == e_analytical_solver::QP_HYBRID)
-        return std::make_unique<QPHybridSolver>();
+        return std::make_unique<QPHybridSolver>(p_placement);
     VPR_FATAL_ERROR(VPR_ERROR_PLACE, "Unrecognized analytical solver type");
     return nullptr;
 }
 
+static inline void populate_update_hybrid_matrix(Eigen::SparseMatrix<double> &A_sparse_diff,
+                                          Eigen::VectorXd &b_x_diff,
+                                          Eigen::VectorXd &b_y_diff,
+                                          const std::vector<double> &diagonal,
+                                          PartialPlacement& p_placement,
+                                          unsigned iteration) {
+    // Aii = Aii+w, bi = bi + wi * xi
+    // TODO: verify would it be better if the initial weights are not part of the function.
+    std::vector<Eigen::Triplet<double>> tripletList;
+    tripletList.reserve(A_sparse_diff.rows());
+    double coeff_pseudo_anchor = 0.01 * std::exp((double)iteration/5);
+    for (int i = 0; i < A_sparse_diff.rows(); i++){
```

Comment:
!!!! I think this is a huge huger error !!!!

A_sparse is composed of movable nodes and star nodes. You should have one anchor point per movable node; however, here it looks like you are doing an anchor point per movable node AND star node. What is most concerning is below where you are using the node locations for i > num_moveable nodes. This is incredibly bad since you are just reading garbage data!

Since you are applying forces to the star node (basically random numbers) it is possible that this is moving your minimum away and causing convergence problems (beyond just being dangerous and incorrect). This may explain why some converges and some do not, and especially why the exponential version does not converge.

This should be `i < p_placement.num_moveable_nodes`. Keep an eye out for this mistake throughout, its an easy slip-up to make.